web3/providers/persistent/persistent.py

```diff
@@ -72,9 +86,63 @@ def __init__(
             self,
             subscription_response_queue_size=subscription_response_queue_size,
         )
+        self._message_listener_task: Optional["asyncio.Task[None]"] = None
+        self._batch_request_counter: Optional[int] = None
+        self._listen_event: asyncio.Event = asyncio.Event()
+        self._max_connection_retries = max_connection_retries
+
         self.request_timeout = request_timeout
         self.silence_listener_task_exceptions = silence_listener_task_exceptions
-        self._max_connection_retries = max_connection_retries
+
+    async def send_func(
+        self, async_w3: "AsyncWeb3", middleware_onion: "MiddlewareOnion"
+    ) -> Callable[..., Coroutine[Any, Any, RPCRequest]]:
+        """
+        Cache the middleware chain for `send`.
+        """
+        middleware = middleware_onion.as_tuple_of_middleware()
+        cache_key = hash(tuple(id(mw) for mw in middleware))
+
+        if cache_key != self._send_func_cache[0]:
+
+            async def send_function(method: RPCEndpoint, params: Any) -> RPCRequest:
+                for mw in middleware:
+                    initialized = mw(async_w3)
+                    method, params = await initialized.async_request_processor(
+                        method, params
+                    )
+
+                return await self.send_request(method, params)
+
+            self._send_func_cache = (cache_key, send_function)
+
+        return self._send_func_cache[1]
```

Comment:
nit: maybe just `return send_function` here?

Reply:
We need to return the cached function here. We don't always build it when we call `send_func` from the manager so this pulls it from the cache instead. It's how the `request_func` is set up as well.

Reply:
Ahh, I see. That makes sense to me. Clever :)